File: 16. Ch-4-2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None
    def insert(self, data):
        if self.root is None:
            self.root = Node(data)
        else:
            self.insertRecursion(self.root, data)
    def insertRecursion(self, node, data):
        if data < node.data:
            if node.left is None:
                node.left = Node(data)
            else:
                self.insertRecursion(node.left, data)
        else:
            if node.right is None:
                node.right = Node(data)
            else:
                self.insertRecursion(node.right, data)

    def lca(self, node, n1, n2):
        while node:
            if n1 < node.data and n2 < node.data:
                node = node.left
            elif n1 > node.data and n2 > node.data:
                node = node.right
            else:
                return node.data

values = list(map(int, input().split()))
n1, n2 = map(int, input().split())

bst = BST()
for v in values:
    if v != -1:
        bst.insert(v)

result = bst.lca(bst.root, n1, n2)
print(f"The Lowest Common Ancestor of {n1} and {n2} is {result}.")

File: 16. Ch-4-2.java
import java.util.*;
class Node {
    int data;
    Node left, right;

    Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BuildTree {
    Node root;
    public void insert(int data) 
    {
        if (root == null)  root = new Node(data);
        else  insertRecursion(root, data);
    }

    public void insertRecursion(Node node, int data) 
    {
        if (data < node.data) 
        {
            if (node.left == null) 
            {
                node.left = new Node(data);
            } 
            else 
            {
                insertRecursion(node.left, data);
            }
        } 
        else 
        {
            if (node.right == null) 
            {
                node.right = new Node(data);
            } 
            else 
            {
                insertRecursion(node.right, data);
            }
        }
    }

    public Node lca(Node root, int n1, int n2) {
        if (root == null) return null;

        if (n1 < root.data && n2 < root.data) {
            return lca(root.left, n1, n2);
        }

        if (n1 > root.data && n2 > root.data) {
            return lca(root.right, n1, n2);
        }

        return root; // split point
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String[] input = sc.nextLine().split(" ");
        BuildTree bst = new BuildTree();

        for (String s : input) {
            int val = Integer.parseInt(s);
            if (val != -1) {
                bst.insert(val);
            }
        }
        int n1 = sc.nextInt();
        int n2 = sc.nextInt();

        Node lcaNode = bst.lca(bst.root, n1, n2);

        System.out.println(
            "The Lowest Common Ancestor of " + n1 + " and " + n2 +
            " is " + lcaNode.data + "."
        );
    }
}
